### DISCUSSION TITLE
[실험] Slot Gate 추가(3진분류 -> 2진분류 2번)

### DISCUSSION THREAD

**Discussion by karim-moon** [bcaitech1/p3-dst-chatting-day#44] [Experiments]
# 문제정의
TRADE논문에 제시되어있는 3진분류(none, don't care, ptr)보다 2진분류 2번을 통해 하나의 게이트가 하나의 문제만을 해결하도록 한다면 모델의 성능이 더욱 올라갈 수 있을것이라 생각하였다.
또한 주어진 데이터에서 value중 가장 많은 클래스가 don't care이며 none과 don't care를 잘 걸러낸다면 모델의 성능과 스코어가 올라갈것이라 예상하였다.

# 해결 아이디어
gate가 늘어나며 코드에서 수정이 되어야하는 부분이 몇군데 있다.

## **data_utils.py**
- **OpenVocabDSTFeatureDoublegat 클래스 생성**
**베이스 라인의 OpenVocabDSTFeature는 gating_id가 1개이므로 None_gating_id, Care_gating_id로 갯수를 늘림**
![data_class](https://user-images.githubusercontent.com/52662993/117566451-a6fae700-b0f1-11eb-897a-17d343c55155.png)

##  **preprocessor.py**
_TRADEPreprocessor 클래스_

- **__init__**
**45가지 쌍의 슬롯 메타가 none인지, none이 아닌지 알려주는 라벨 벡터와 don't care인지 don'care가 아닌지 구분하는 벡터가 
모델의 input으로 들어가기때문에 gating2id를 추가해주었다.**
![gating](https://user-images.githubusercontent.com/52662993/117565855-79f90500-b0ee-11eb-81dd-74b74d11321f.png)
**마찬가지로 예측값을 변환할 때 각 value를 통해 변환되므로 id2gating역시 두가지로 나누었다.**
![id2gating](https://user-images.githubusercontent.com/52662993/117566248-78c8d780-b0f0-11eb-8696-41954145d25c.png)

- **_convert_example_to_feature**
**gating2id를 거진 값들이 45개 원소를 가진 벡터형태로 gating_ids가 만들어지기 때문에 gating_id역시 나누었다.**
![make_label](https://user-images.githubusercontent.com/52662993/117566300-d5c48d80-b0f0-11eb-9c1f-476ad0431f30.png)
**return부분에 기존 사용되던 OpenVocabDSTFeature 대신 OpenVocabDSTFeatureDoublegat 데이터 클래스에 선언**
![return](https://user-images.githubusercontent.com/52662993/117566518-fccf8f00-b0f1-11eb-83f1-c4140982737c.png)

- **recover_state**
**기존 gate_list만 받았던 함수를 none을 분류하는 none_list, don't care를 분류하는 care_list를 입력받도록 수정**
![recover_state_input](https://user-images.githubusercontent.com/52662993/117566653-d1996f80-b0f2-11eb-968d-edb1c37a3c59.png)
**예측치를 텍스트 결과로 변환시 하나의 게이트 입력으로 처리하는 것이 아닌 None_id2gating, Care_id2gating으로 나누어 처리**
![value_gating](https://user-images.githubusercontent.com/52662993/117566665-f42b8880-b0f2-11eb-99e3-364c75c315f9.png)

- **collate_fn**
**마찬가지로 배치단위로 모델에 두개의 게이트값을 주어야하기 때문에 None_gating_ids, Care_gating_ids를 추가하였다.**
![add](https://user-images.githubusercontent.com/52662993/117566766-93e91680-b0f3-11eb-845a-0fd06c490c2c.png)
**return 역시 수정!!**
![return_collate](https://user-images.githubusercontent.com/52662993/117566774-a1060580-b0f3-11eb-88e2-6fc637b4306a.png)

 ## **model.py**
_SlotGenerator 클래스_

- **__init__**
**3개의 클래스(none, ptr, don't care)로 구분하는 Linear layer를 두개로 나누어 각각 2진분류를 하도록 분리하였다.**
![w_gen](https://user-images.githubusercontent.com/52662993/117566894-33a6a480-b0f4-11eb-877d-7a11c4d1e702.png)

- **forward**
**gating이 일어나는 경우는 전체 컨텍스트 벡터와 decoder가 계산될 때( k == 0 )이므로 역시 두개의 분리된 Linear layer를 거치고 logit을 만든다.
view(batch_size, J(슬롯페어 쌍 갯수 = 45), 게이트의 경우의수(none, not_none or don't care, others) = 2)이므로 기존 args.n_gate를 2로 수정하였다.**
!!!return 역시 none_gate_output, care_gate_output으로 수정된다.
![k_equal_zero](https://user-images.githubusercontent.com/52662993/117567048-1a522800-b0f5-11eb-8cd6-e3d5f3a3fb2f.png)

## **train.py**

- **loss 클래스 생성**
none 분류 gate와 don't care 분류 gate가 서로 다른 값을 갖기때문에 loss function을 두가지로 분류하였다. (CrossEntropyLoss)
![calc_loss](https://user-images.githubusercontent.com/52662993/117568872-ca2b9380-b0fd-11eb-8858-d1165a6bc62a.png)

- **loss 계산**
loss가 계산되는 부분 역시 두가지로 나뉘는 부분과 최종 계산시 loss_gen, loss_none, loss_care를 더해주는 형식으로 진행하였다.
![loss_gate](https://user-images.githubusercontent.com/52662993/117568982-4cb45300-b0fe-11eb-847e-3e5e121738b8.png)
![sumation](https://user-images.githubusercontent.com/52662993/117568993-550c8e00-b0fe-11eb-8901-a4786392ff29.png)

## **inference.py**

- **inference**
batch에서 none_ids, care_ids가 포함되어있기때문에 가장 확률이 높은 값을 max로 뽑아내어 각각 list로 만들고 이를 통해 gating이 이루어진다.
![inference](https://user-images.githubusercontent.com/52662993/117569110-f1cf2b80-b0fe-11eb-88e4-89b2495c9717.png)

# 진행상황
#16 을 참조해주세요!!

변경사항 : 3개의 gate를 2개로 줄일 때 인코더를 bert를 사용

# 결과
- 기준(baseline) 
![baseline](https://user-images.githubusercontent.com/52662993/117569204-82a60700-b0ff-11eb-8be5-3cfa38c56ccd.png)

- 3 gate(none, don't care, ptr) score
오히려 베이스라인보다 점수가 떨어진걸 확인할 수 있다.
생각해보면 정답파일은 none이 제외된, don't care와 ptr로 이루어진 파일일 것이다. 만약 베이스라인 모델이 이미 none을 충분히 잘 걸러낼 수 있는 모델이라면 none을 잘 걸러내는 일은 무의미할 수 있을것이다.
![tripple_gate_score](https://user-images.githubusercontent.com/52662993/117569242-b1bc7880-b0ff-11eb-8c14-23b7eb2e6a6a.png)

- 2gate(none, don't care)
게이트를 하나 줄였기때문에 성능이 올라갔다기보단 GRUEncoder를 BERT로 교체하였다.
성능 향상은 bert 인코더 때문일것으로 추측된다.
3개의 게이트를 2개로 줄였지만 사실 ptr게이트는 inference에 영향을 미치지 않고 layer또한 따로 존재하기때문에 3gate와는 별반 다를것이 없을것이다.
![bert](https://user-images.githubusercontent.com/52662993/117569284-f5af7d80-b0ff-11eb-886f-2276fde2d501.png)

# 후속 실험
Special mission에서도 나와있듯이 yes와 no를 구분하는 게이트를 분리하여 진행을 해볼 수 있을것이다.
